### PR TITLE
refactor: Plugin cleanup

### PR DESCRIPTION
**Pull request by mofojed**
- Clean up a bunch of the typings around plugins
- Add hooks for panel registration, like `useComponent`
  - Removes a lot of the boilerplate code currently used in plugins
- One hydrate method now is used for new panels and rehydrating dehydrated panels alike
- Makes the examples for https://github.com/deephaven/deephaven-js-plugins/issues/12 much cleaner

### CODE REVIEW

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/code-studio/src/main/AppMainContainer.tsx [hunk 1]
```diff
@@ -709,55 +706,69 @@ export class AppMainContainer extends Component<
...
+  ): DehydratedDashboardPanelProps & {
     getDownloadWorker: () => Promise<ServiceWorker>;
     loadPlugin: (pluginName: string) => TablePlugin;
     localDashboardId: string;
     makeModel: () => Promise<IrisGridModel>;
   } {
     const { connection } = this.props;
+    let metadata: IrisGridPanelMetadata;
+    // TODO: Why is this showing a TS error in eslint in VSCode??
+    if (isIrisGridPanelMetadata(props.metadata)) {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/main/AppMainContainer.tsx:735]: Yea I'm not seeing the error anymore

File: packages/code-studio/src/main/AppMainContainer.tsx [hunk 2]
```diff
@@ -709,55 +706,69 @@ export class AppMainContainer extends Component<
...
+      };
+    }
+
     return {
       ...props,
       getDownloadWorker: DownloadServiceWorkerUtils.getServiceWorker,
       loadPlugin: this.handleLoadTablePlugin,
       localDashboardId: id,
-      makeModel: () => createGridModel(dh, connection, props.metadata, type),
+      makeModel: () => createGridModel(dh, connection, metadata),
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/main/AppMainContainer.tsx:749]: Yea true, that is strange. I'll add an `else` on the above if/else statement.

File: packages/code-studio/src/main/AppMainContainer.tsx [hunk 3]
```diff
@@ -691,11 +690,9 @@ export class AppMainContainer extends Component<
   }
 
   hydrateDefault(
-    props: {
-      metadata?: { type?: string; id?: string; name?: string };
-    } & PanelProps,
+    props: DehydratedDashboardPanelProps,
     id: string
-  ): DashboardPanelProps & { fetch?: () => Promise<unknown> } {
+  ): DehydratedDashboardPanelProps & { fetch?: () => Promise<unknown> } {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/code-studio/src/main/AppMainContainer.tsx:695]: Hmm well they don't have the `GLPanelProps` set (`gleventHub`,`glContainer`), those get set by `golden-layout` on mount: https://github.com/mofojed/web-client-ui/blob/218cf5224bcba3734025749fe55b1161bc3bb4a6/packages/golden-layout/src/utils/ReactComponentHandler.tsx#L166

There is still the added hydration properties ( the `& { fetch?: () => Promise<unknown> }`) here and in the below functions. And then there are the props provided by Redux... defined within IrisGridPanel itself...
In short, you're right, I should clean this up better.

File: packages/dashboard/src/layout/hooks.ts [hunk 4]
```diff
@@ -1,12 +1,26 @@
...
+import {
+  DashboardPanelProps,
+  DashboardPluginComponentProps,
+  DehydratedDashboardPanelProps,
+  PanelComponentType,
+  PanelDehydrateFunction,
+  PanelHydrateFunction,
+} from '../DashboardPlugin';
+import PanelEvent, { PanelOpenEventDetail } from '../PanelEvent';
+import { LayoutUtils } from '.';
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard/src/layout/hooks.ts:16]: Bigtime

File: packages/dashboard/src/DashboardLayout.tsx [hunk 5]
```diff
@@ -133,6 +133,8 @@ export function DashboardLayout({
           <PanelErrorBoundary glContainer={glContainer} glEventHub={glEventHub}>
             {/* eslint-disable-next-line react/jsx-props-no-spreading */}
             <PanelWrapperType {...props}>
+              {/* TODO: Verify the id is getting populated from somewhere */}
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard/src/DashboardLayout.tsx:136]: Yea indeed there was a misalignment with how the panel ID was provided. I have it as a required prop in `PanelProps`, but it only needs to be set on the `ItemConfig`; from within the panel you can get the ID using `LayoutUtils.getIdFromContainer`.

I'm cleaning this up, but I'm also thinking about how we can make this simpler. Particularly in a programmatic layout world, where we may allow nesting of "panels" (e.g. multiple charts/tables in one "panel").

File: packages/dashboard-core-plugins/src/panels/IrisGridPanel.tsx [hunk 6]
```diff
@@ -209,6 +204,20 @@ interface IrisGridPanelState {
   gridStateOverrides: Partial<GridState>;
 }
 
+function getTableName(metadata?: PanelMetadata): string {
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard-core-plugins/src/panels/IrisGridPanel.tsx:207]: Yea union w/ undefined is clearer.

File: packages/dashboard-core-plugins/src/panels/IrisGridPanel.tsx [hunk 7]
```diff
@@ -712,8 +721,8 @@ export class IrisGridPanel extends PureComponent<
     this.setState(
       () => null,
       () => {
-        const { glEventHub, inputFilters, metadata } = this.props;
-        const { table } = metadata;
+        const { glEventHub, inputFilters } = this.props;
+        const table = this.getTableName();
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard-core-plugins/src/panels/IrisGridPanel.tsx:725]: No, there's a `this.getTableName` that takes the instance metadata and calls the static `getTableName` with metadata. I've renamed the static one.

File: packages/dashboard-core-plugins/src/events/IrisGridEvent.ts [hunk 8]
```diff
@@ -1,4 +1,12 @@
-class IrisGridEvent {
+import { RowDataMap } from '@deephaven/jsapi-utils';
+import { IrisGridPanel } from '../panels/IrisGridPanel';
+
+export type IrisGridDataSelectedEventCallback = (
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard-core-plugins/src/events/IrisGridEvent.ts:4]: Yea... I can remove, but there should be a better way to define what gets passed when an event is emitted.

File: packages/dashboard-core-plugins/src/ConsolePlugin.tsx [hunk 9]
```diff
@@ -534,7 +533,10 @@ export function ConsolePlugin(
...
 
   const hydrateNotebook = useCallback(
-    (panelProps: PanelProps, panelDashboardId: string): DashboardPanelProps =>
+    (
+      panelProps: DehydratedDashboardPanelProps,
+      panelDashboardId: string
+    ): DehydratedDashboardPanelProps =>
       DashboardUtils.hydrate(
         {
           ...panelProps,
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard-core-plugins/src/ConsolePlugin.tsx:542]: Should be the `usePanelRegistration` hook that @bmingles created, not the one I created. These panels open based on a different event occurring, not when an object of a specific type gets created.

File: packages/dashboard/src/layout/useDashboardPanel.ts [hunk 10]
```diff
@@ -56,16 +58,15 @@ export function useComponent<
...
-        type === variableName;
+        (Array.isArray(supportedTypes) && supportedTypes.includes(type)) ||
+        type === supportedTypes;
       if (!isSupportedType) {
         // Only want to listen for your custom variable types
         return;
       }
       const metadata = { id: widgetId, name, type };
       let props: DehydratedDashboardPanelProps & { fetch?: typeof fetch } = {
         localDashboardId: id,
```
Comment (mattrunyon) [inline packages/dashboard/src/layout/useDashboardPanel.ts:69]: What is the "local" for in this variable name? Why not just `dashboardId`? It's just the ID of the dashboard the panel is mounted in, right? So in community the same for all panels, but in enterprise different for each code studio/dashboard. "local" seems like an unneeded qualifier here (what other types of `dashboardId` are there?)

Might also make sense in a context (but this one might be a bigger refactor than `registerComponent`). Maybe they go in the same context and you can have some access hook like `const { id: dashboardId, registerComponent } = useDashboardInfo();`. Just a thought that might simplify a few things or clear up some uses.

If we changed to a context, then probably should keep the props and deprecate if enterprise still needs the props.

File: packages/dashboard/src/layout/useListener.ts [hunk 11]
```diff
@@ -0,0 +1,28 @@
...
+import type { EventEmitter } from '@deephaven/golden-layout';
+
+/**
+ * Listen for a specific event on an EventEmitter
+ * @param eventEmitter EventEmitter to listen to
+ * @param eventName Name of the event to listen to
+ * @param callback Callback to call when the event is triggered
+ */
+export function useListener(
+  eventEmitter: EventEmitter,
```
Comment (mattrunyon) [inline packages/dashboard/src/layout/useListener.ts:11]: Could this be fetched from a context too? Not trying to overuse contexts here, but a lot of this feels like the right use for them. We have hooks being used in components which need the state from a few levels up in the tree. And a lot of these are just using the exact same pattern. Pretty much all calls to this hook are either `useListener(goldenLayout, ...)` or `useListener(goldenLayout.eventHub, ...)`.

Maybe a `useGoldenLayoutListener` and `useListener`? The cases with `useListener(goldenLayout` are specifically for golden layout events like panel picked up or panel dropped. The others are arbitrary events we define and just plumb through golden layout

**Review by mattrunyon — CHANGES_REQUESTED** [review]

File: packages/dashboard-core-plugins/src/ConsolePlugin.tsx [hunk 12]
```diff
@@ -547,26 +548,11 @@ export function ConsolePlugin(
...
-        ),
-      ];
-
-      return () => {
-        cleanups.forEach(cleanup => cleanup());
-      };
-    },
-    [registerComponent, hydrateConsole, hydrateNotebook]
-  );
+  usePanelRegistration(registerComponent, ConsolePanel, hydrateConsole);
```
[the remark this answers is not in the document]
Comment (mofojed) [reply packages/dashboard-core-plugins/src/ConsolePlugin.tsx:551]: Yea it could be, along with everything else that's passed in via `cloneElement` in [DashboardLayout](https://github.com/deephaven/web-client-ui/blob/3247d274dd724c5c65caa7047ad843c03d42aee4/packages/dashboard/src/DashboardLayout.tsx#L320) (`panelManager`, `layout`, `id`)

File: packages/dashboard/src/layout/useDashboardPanel.ts (hunk 10 above)
Comment (mofojed) [reply packages/dashboard/src/layout/useDashboardPanel.ts:69]: Yea there's no real point to having the `local` bit. I think there was something where we needed to create an ID for a dashboard in advance when we hadn't saved that dashboard yet, but no reason to name it `local`

File: packages/dashboard/src/layout/useListener.ts (hunk 11 above)
Comment (mofojed) [reply packages/dashboard/src/layout/useListener.ts:11]: Yup context would make sense here

Comment (mofojed) [reply packages/dashboard/src/layout/useListener.ts:11]: We'll address contexts in a follow-up PR, this PR shouldn't be breaking and changing to contexts would be.

File: packages/code-studio/src/main/AppMainContainer.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (mattrunyon) [reply packages/code-studio/src/main/AppMainContainer.tsx:695]: These can be cleaned up in a contexts + single shared hydration function for each panel that just gets metadata + variable info PR

**Review by mattrunyon — APPROVED** [review]

**Review by mattrunyon — APPROVED** [review]
